### PR TITLE
Don't accept unknown command line arguments

### PR DESCRIPTION
**Pull request by dcbaker**
Mostly because I can neither type nor spell.

### CODE REVIEW

File: mesonbuild/optinterpreter.py [hunk 1]
```diff
@@ -128,6 +128,11 @@ def process(self, option_file):
                 e.colno = cur.colno
                 e.file = os.path.join('meson_options.txt')
                 raise e
+            bad = [o for o in self.cmd_line_options if o not in self.options]
+            if bad:
+                sub = 'In subproject {}: '.format(self.subproject) if self.subproject else ''
+                raise OptionException(
+                    '{}Unknown command line options: "{}"'.format(sub, bad))
```
Comment (keszybz) [inline mesonbuild/optinterpreter.py]: I think you should make explicit formatting here, `', '.join(bad)` or so.

File: run_unittests.py [hunk 2]
```diff
@@ -1681,6 +1682,23 @@ def test_boost_BOOST_ROOT_dependency(self):
         self.assertMesonRaises("dependency('boost')",
                                "(BOOST_ROOT.*absolute|{})".format(self.dnf))
 
+    def test_bad_option(self):
+        tdir = os.path.join(self.unit_test_dir, '18 bad command line options')
+        os.environ['MESON_FORCE_BACKTRACE'] = '1'
+        self.init(tdir, extra_args="-Dopt=bar --prefix=/tmp", inprocess=True)
+        self.wipe()
+        with self.assertRaisesRegex(OptionException, r'Unknown command line options:.*foo.*'):
```
Comment (keszybz) [inline run_unittests.py]: ... and test the exact form here.

**Comment by dcbaker** [comment]
I'm having a bad morning, so if I come off grumpy, I am, just not with you :)

I strongly disagree about making it a warning. It's not useful as a developer or as an end user to think that I've configured the project a specific way, and then find it that I didn't. There have been a number of times that that exact situation has led me to push patches as "working" when in fact I tested the default configuration in both cases, and the new case I was trying to test was completely broken. Or when an option has changed (or changed form) and I tried to pass the old version, I'd much rather get a hard error than a warning I almost certainly wont notice.

I mean, I know that you work on systemd, I work on mesa, and both projects create more than one terminal height's worth of data. The warning would be at the very top of the terminal output for `meson builddir`, would you honestly notice it, because I wouldn't.

**Comment by NickeZ** [comment]
I agree with dcbaker, the boost build system "b2" only warns you if you mistype something and then continues to build something I didn't ask for.. that is not a reasonable behavior.

**Comment by dcbaker** [comment]
The latest code resolves the specific suggestions from @keszybz (though not the warning/error concern), I'd like to see some other people weigh in on that.

It also fixes test errors, commits some new files that were missing, fixes error message if multiple bad options are passed, fixes passing meson reserved options (like c_*).

**Comment by jpakkane** [comment]
In these cases we have often made the breaking change only a warning at first, with a message saying this will become a hard error in the future. This gives people some time to clean up old code that may depend on the old behaviour.

**Comment by dcbaker** [comment]
Updated to warn instead of fail if bad options are passed.

File: run_unittests.py [hunk 3]
```diff
@@ -38,6 +37,7 @@
 from mesonbuild.environment import Environment
 from mesonbuild.dependencies import DependencyException
 from mesonbuild.dependencies import PkgConfigDependency, ExternalProgram
+from mesonbuild.optinterpreter import OptionException
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

'mesonbuild.optinterpreter.OptionException' imported but unused

**Comment by jpakkane** [comment]
Test failure seems relevant.

**Comment by jpakkane** [comment]
ping

**Comment by dcbaker** [comment]
Yup, I was looking at this today and I think I can fix it. I'll be pretty busy tomorrow but I plan to have an updated pull request on monday.

**Comment by dcbaker** [comment]
@jpakkane I think this is ready now.

**Pull request merged**